lib/OpenQA/Scheduler/Scheduler.pm

```diff
@@ -138,17 +150,397 @@ sub _prefer_parallel {
+                    #Save it, in next round we will check if we got answer from worker and see what to do
+                    $ws_allocated_jobs->{$allocated->{id}}
+                      = {assigned_worker_id => $allocated->{assigned_worker_id}, result => $allocated->{result}};
+                }
+                else {
+                    die "Failed rollback of job" unless $job->reschedule_rollback($worker);
+                }
+            }
+            catch {
+                log_debug("Failed to set worker in scheduling state :( bummer! Reason: $_");
+                $failure++
+                  if OpenQA::Scheduler::CONGESTION_CONTROL()
+                  && OpenQA::Scheduler::BUSY_BACKOFF();    # double it, we might be in a heavy load condition.
+                                                           # Also, the workers could be massively in re-registration.
+            };
+
+        }
+        else {
+            log_debug("Failed sending job '"
+                  . $allocated->{id}
+                  . "' to worker '"
+                  . $allocated->{assigned_worker_id} . "' : "
+                  . pp($res));
+            $failure++
+              if OpenQA::Scheduler::CONGESTION_CONTROL()
+              && OpenQA::Scheduler::BUSY_BACKOFF();        # We failed dispatching it. We might be under load
+
+            try {
+                $worker->unprepare_for_work;
+            }
+            catch {
+                log_debug("Failed resetting unprepare worker :( bummer! Reason: $_");
+                $failure++
+                  if OpenQA::Scheduler::CONGESTION_CONTROL()
+                  && OpenQA::Scheduler::BUSY_BACKOFF();    # double it, we might be in a heavy load condition.
+                                                           # Also, the workers could be massively in re-registration.
+            };
+
+
+            # put the job in scheduled state again.
+            # we could search, dispatch the job and then update database in one transactions.
+            # but it's better avoid to put ipc communication over dbus inside a transaction
+            # (caused more-than-one complete sqlite database breakage locally)
+            try {
+                die "Failed reset" unless $job->reschedule_state;
+            }
+            catch {
+                log_debug("Failed resetting job '$allocated->{id}' to scheduled state :( bummer! Reason: $_");
+                $failure++
+                  if OpenQA::Scheduler::CONGESTION_CONTROL()
+                  && OpenQA::Scheduler::BUSY_BACKOFF();    # double it, we might be in a heavy load condition.
+                                                           # Also, the workers could be massively in re-registration.
+            };
+        }
+    }
+
+    $failure--
+      if $failure > 0
+      && $successfully_allocated > 0
+      && OpenQA::Scheduler::CONGESTION_CONTROL()
+      && OpenQA::Scheduler::BUSY_BACKOFF();
```

Comment:
this was meant to be `(OpenQA::Scheduler::CONGESTION_CONTROL() || OpenQA::Scheduler::BUSY_BACKOFF()) `